### ISSUE TITLE
Question about client threading  

### ISSUE THREAD

**Issue by ali-z-can** [warmcat/libwebsockets#2071]
Hi, 
This might be a trivial question but please cope with me, I'm a newbie in websockets and C. 

I have a Go websocket server which pumps data over a Gbps connection. I have implemented a websocket client which just reads the incoming packet lengths and it works fine on my pc. The target environment has limited resources(an embedded environment) and can not catch up with the Gbps connection speed. I thought having 1 thread just for event loop and two threads for callback functions might do the trick. I know lws offers vhost's but I can not figure out how to use them for client reads. I would appreciate any suggestions, resources or examples. 

Thanks in advance.

P.S. Target machine is capable of handling Gpbs connections(tested with iperf3).

**Comment by lws-team** [comment]
I think you should look closer at the actual problem before throwing threads on it.

Does the client do a lot of work for each rx message?  Is the cpu at 100% dealing with it?  If not, threads won't help.

What is the size of the messages the server is sending?

Profile where the client spends its time?

Vhosts for client aren't really relevant, they are just a place to hang the client SSL_CTX.

**Comment by ali-z-can** [comment]
Thanks for your answer Andy,

Client only adds up total message sizes, so i don't think thats a lot of work.
CPU is at 75-85% range during execution(I guess this means threads won't help much). 
Server sends messages of 65536 bytes. (Starting with 8192 bytes and go up to 6553). 
Client spends most of its time(33%) at lws_ws_client_rx_sm.(Profiled with perf) 
I'm also monitoring in which speed client receives messages, it's about 110Mbps, 9 times lower than expected.

**Comment by lws-team** [comment]
> 9 times lower than expected.

Yeah but you don't have any particular reason for "expecting" end-end throughput of 1gbps once it's doing some work.

tcp will cause it to adapt to the rate the rx packets are being consumed at the receiver.

Check you don't have permessage-deflate negotiated.

Lws chooses a modest default rx size because it works on RTOS type clients, you can adjust it higher to get more throughput

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-context-vhost.h#n616

and in the protocol you can set the rx buffer size

https://libwebsockets.org/git/libwebsockets/tree/include/libwebsockets/lws-protocols-plugins.h#n57

**Comment by ali-z-can** [comment]
I know I could not get Gbps speed at application level but I'm trying to implement sort of a speed test and since I'm not dealing with the payload I'm expecting to get something at least close to connection speed. (850-900 Mbps maybe ?)

I checked that I don't have permessage-deflate extention. (Looked up from wireshark I hope its a correct way to do so.)

I set rx buffer size to 524288 bytes at lws_protocols.

Also in wireshark I get a ton of TCP zero window (Don't know if it is relevant though).

**Comment by lws-team** [comment]
Did you set the context info struct

```
        unsigned int pt_serv_buf_size;
        /**< CONTEXT: 0 = default of 4096.  This buffer is used by
         * various service related features including file serving, it
         * defines the max chunk of file that can be sent at once.
         * At the risk of lws having to buffer failed large sends, it
         * can be increased to, eg, 128KiB to improve throughput. */
```

because the data is in there before the protocol rx buffer.

**Comment by ali-z-can** [comment]
I haven't, thanks that helped a little. I increased pt_serv_buf_size (tried 64KB up to 64MB) but the speed only went from 110 Mbps to 130 Mbps. Also TCP zero windows are still existent. I'm really sorry that I'm wasting your time Andy, but i feel kinda lost. Now I know neither tcp buf nor protocol buf sizes are not the issue. Also since my cpu usage is still not at 100% (but its increased to 95-98%) throwing threads at it won't fix right? Is there any other possible reason for getting that low of results that you think of?

**Comment by lws-team** [comment]
Build with -DCMAKE_BUILD_TYPE=RELEASE to disable the debug logging might do something.

You have the problem so in the end you have to debug where it's spending its time.

**Comment by ali-z-can** [comment]
I was building with -DCMAKE_BUILD_TYPE=RELEASE.

So, is there way to have three threads as fallows; 
-First thread: Only handles event loop.
-Second and third threads: Only handles callbacks from event loop.

**Comment by lws-team** [comment]
Sure that is what threadpool is.

Are you doing work with the rx data at the moment?  What happens if you just ignore it and just return from the RX callback.

If you are not doing any work with it, clearly throwing threads and locking into the mix is only going to make things worse.

**Comment by ali-z-can** [comment]
Not any work exactly with the rx data, I'm just using rx data length. I just tried doing nothing and returning from RX callback, nothing changes still too slow. My thought process was like this; right now one thread handling incoming data and it's callback, even though my callback is nothing fancy(just incrementing a global variable) the thread needs to hold on to the data longer(since as I know there is no way droping rx data and just pass rx data length to callback functions). So having a thread dedicated to handling callback might help.

**Comment by lws-team** [comment]
I took another look at this, the reason that client ws bulk data handling is not as efficient is that unlike the server ws path, the client ws path handles the bulk data bytewise.  In the server path, it handles the ws headers bytewise, but then understands that the next xxx bytes are payload, and handles them blockwise.

It's possible to align the client ws path to what the server does

Server: https://libwebsockets.org/git/libwebsockets/tree/lib/roles/ws/server-ws.c#n1043-1093
             https://libwebsockets.org/git/libwebsockets/tree/lib/roles/ws/server-ws.c#n805
Client: https://libwebsockets.org/git/libwebsockets/tree/lib/roles/ws/client-parser-ws.c#n336

as you can see it's not so trivial dealing with the masking, permessage-deflate and accounting correctly.  But the server way is there to show how to do it.
